### PR TITLE
wgengine/tsdns: replace connections when net link changes (macOS)

### PR DESCRIPTION
**Pull request by josharian**
See individual commit message for details.

Tested manually on macOS. If we're happy with this direction, we can expand to other platforms.

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/788)

### CODE REVIEW

File: wgengine/tsdns/forwarder.go [hunk 1]
```diff
@@ -285,3 +272,189 @@ func (f *forwarder) forward(query Packet) error {
 
 	return nil
 }
+
+// A fwdconn manages a single connection used to forward DNS requests.
+// Net link changes can cause a *net.UDPConn to become permanently unusable, particularly on macOS.
+// fwdconn detects such situations and transparently creates new connections.
+type fwdconn struct {
```
Comment (bradfitz) [inline wgengine/tsdns/forwarder.go:279]: fwdConn?

File: wgengine/tsdns/neterr_darwin.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+func isSyscallErrno(err error, errno syscall.Errno) bool {
+	var n syscall.Errno
+	if !errors.As(err, &n) {
+		return false
+	}
+	return n == errno
+}
+
+func networkIsDown(err error) bool {
+	return isSyscallErrno(err, syscall.ENETDOWN)
```
Comment (bradfitz) [inline wgengine/tsdns/neterr_darwin.go:21]: why not errors.Is? you can prevent the allocation into the target interface if you do it once in a global variable:

```go
  var errSyscallENETDONE error = syscall.ENETDOWN
```

?

And below.

File: wgengine/tsdns/forwarder.go [hunk 3]
```diff
@@ -285,3 +272,189 @@ func (f *forwarder) forward(query Packet) error {
...
+			return
+		}
+		if errors.Is(err, os.ErrDeadlineExceeded) {
+			// We intentionally closed this connection.
+			// It has been replaced by a new connection. Try again.
+			continue
+		}
+		// Something else went wrong.
+		// We have three choices here: try again, give up, or create a new connection.
+		var operr *net.OpError
```
Comment (bradfitz) [inline wgengine/tsdns/forwarder.go:341]: opErr ?

File: wgengine/tsdns/forwarder.go [hunk 4]
```diff
@@ -285,3 +272,189 @@ func (f *forwarder) forward(query Packet) error {
...
+	c.logf = logger.WithPrefix(logf, fmt.Sprintf("fwdconn %d: ", idx))
+	c.change = sync.NewCond(&c.mu)
+	// c.conn is created lazily in send
+	return c
+}
+
+// send sends packet to dst using fwdconn's connection.
+// It is best effort. It is UDP, after all. Failures are logged.
+func (c *fwdconn) send(packet []byte, dst net.Addr) {
+	b := backoff.NewBackoff("tsdns-fwdconn-send", c.logf, 30*time.Second)
```
Comment (bradfitz) [inline wgengine/tsdns/forwarder.go:309]: the common case is success. can you lazily allocate this?

File: wgengine/tsdns/forwarder.go [hunk 5]
```diff
@@ -285,3 +272,189 @@ func (f *forwarder) forward(query Packet) error {
...
+type fwdconn struct {
+	// logf allows a fwdconn to log
+	logf logger.Logf
+
+	// wg tracks the number of outstanding conn.Read and conn.Write calls
+	wg sync.WaitGroup
+	// change allows calls to read to block until a the network connection has been replaced
+	change *sync.Cond
+
+	// mu protects fields that follow it; it is is also change's Locker
```
Comment (bradfitz) [inline wgengine/tsdns/forwarder.go:288]: missing periods at the end of all these sentences

File: wgengine/tsdns/forwarder.go [hunk 6]
```diff
@@ -285,3 +272,189 @@ func (f *forwarder) forward(query Packet) error {
...
+	mu sync.Mutex
+
+	// closed tracks whether fwdconn has been permanently closed
+	closed bool
+
+	// conn is the current active connection
+	conn *net.UDPConn
+}
+
+func newFwdConn(logf logger.Logf, idx int) *fwdconn {
```
Comment (bradfitz) [inline wgengine/tsdns/forwarder.go:298]: this constructor name's case is more evidence the type should be fwdConn :)

**Review by bradfitz — APPROVED** [review]
Congrats on touching MagicDNS last!

File: wgengine/tsdns/neterr_darwin.go (hunk 2 above)
Comment (josharian) [reply wgengine/tsdns/neterr_darwin.go:21]: Nice, thanks.

File: wgengine/tsdns/forwarder.go (hunk 4 above)
Comment (josharian) [reply wgengine/tsdns/forwarder.go:309]: It doesn't move the benchmarks, but since it's done now, I'll leave it in place.

**Comment by josharian** [comment]
All done, thanks. PTAL.

**Comment by josharian** [comment]
Oh n/m, easy to miss the approval. No need to take another look. :)

**Pull request merged**